### PR TITLE
CRI: Allow configuring stdout/stderr streams for Exec/Attach requests

### LINKED ISSUE #44448: CRI: exec/attach requests don't let you disable stdout or stderr
I'm working on rebasing OpenShift on top of Kubernetes 1.6 and I just noticed that if you try to exec/attach to a pod, the CRI protobuf messages for AttachRequest and ExecRequest only allow you to specify stdin and tty. You are not able to specify stdout or stderr. The handlers for serving exec and attach streaming requests force stdout to true, and stderr to true when not using a tty. Is there a reason that stdout and stderr are hard-coded and not optional?

I'm able to work around this by saying I "want" stdout and stderr, and sending their data to `ioutil.Discard`, but I don't think that's a good long-term solution. I believe this is a regression and we should restore the ability to selectively enable both stdout and stderr.

Thoughts?

@yujuhong @dchen1107 @vishh @Random-Liu @timstclair @smarterclayton @derekwaynecarr @liggitt @sttts @kubernetes/sig-node-bugs

### PR DESCRIPTION
**Pull request by yujuhong**
Add stdout/stderr to exec and attach requests. Also check the request to
ensure it meets the requirements.

**Which issue this PR fixes**: fixes #44448

```release-note
CRI: Add stdout/stderr fields to Exec and Attach requests.
```

### CODE REVIEW

File: pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto [hunk 1]
```diff
@@ -827,7 +827,17 @@ message ExecRequest {
...
     bool tty = 3;
     // Whether to stream stdin.
+    // One of `stdin`, `stdout`, and `stderr` MUST be true.
     bool stdin = 4;
+    // Whether to stream stdout.
+    // One of `stdin`, `stdout`, and `stderr` MUST be true.
+    bool stdout = 5;
+    // Whether to stream stderr.
+    // One of `stdin`, `stdout`, and `stderr` MUST be true.
+    // If `tty` is true, `stderr` MUST be false. Multiplexiing is not supported
```
Comment (mrunalp) [inline pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto]: typo: Multiplexiing

File: pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto [hunk 2]
```diff
@@ -839,10 +849,20 @@ message AttachRequest {
...
     bool stdin = 2;
     // Whether the process being attached is running in a TTY.
     // This must match the TTY setting in the ContainerConfig.
     bool tty = 3;
+    // Whether to stream stdout.
+    // One of `stdin`, `stdout`, and `stderr` MUST be true.
+    bool stdout = 4;
+    // Whether to stream stderr.
+    // One of `stdin`, `stdout`, and `stderr` MUST be true.
+    // If `tty` is true, `stderr` MUST be false. Multiplexiing is not supported
```
Comment (mrunalp) [inline pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto]: same typo here.

File: pkg/kubelet/server/streaming/server_test.go [hunk 3]
```diff
@@ -78,104 +67,159 @@ func TestGetExec(t *testing.T) {
...
-		assert.Equal(t, test.cmd, req.(*runtimeapi.ExecRequest).Cmd, "testcase=%+v", test)
-		assert.Equal(t, test.tty, req.(*runtimeapi.ExecRequest).Tty, "testcase=%+v", test)
-		assert.Equal(t, test.stdin, req.(*runtimeapi.ExecRequest).Stdin, "testcase=%+v", test)
+		require.True(t, ok, "token %s not found!", token)
+		assert.Equal(t, expectedReq, req)
+	}
+	request := &runtimeapi.ExecRequest{
+		ContainerId: testContainerID,
+		Cmd:         []string{"echo", "foo"},
+		Tty:         true,
```
Comment (ncdc) [inline pkg/kubelet/server/streaming/server_test.go]: Would it be worth testing true & false values for tty, stdin, stdout, and stderr (for the valid combinations)? Or do you think the validate test is sufficient?

File: pkg/kubelet/server/streaming/server_test.go [hunk 4]
```diff
@@ -260,21 +304,24 @@ func TestServePortForward(t *testing.T) {
 	doClientStreams(t, "portforward", stream, stream, nil)
 }
 
+//
```
Comment (ncdc) [inline pkg/kubelet/server/streaming/server_test.go]: ?

File: pkg/kubelet/server/streaming/server_test.go (hunk 3 above)
Comment (yujuhong) [reply pkg/kubelet/server/streaming/server_test.go]: I think the validate tests should be sufficient, but wouldn't mind adding more test cases here if you prefer

File: pkg/kubelet/server/streaming/server_test.go (hunk 4 above)
Comment (yujuhong) [reply pkg/kubelet/server/streaming/server_test.go]: Artifact left from debugging. Will remove

**Comment by yujuhong** [comment]
This still assumes that if `tty` is set, we'll combine both `stdout` and `stderr` streams into `stdout`.
Are folks okay with that? @ncdc @tallclair @kubernetes/sig-node-pr-reviews

**Comment by ncdc** [comment]
> This still assumes that if tty is set, we'll combine both stdout and stderr streams into stdout.
Are folks okay with that?

There is no other option.

File: pkg/kubelet/server/streaming/server_test.go (hunk 3 above)
Comment (ncdc) [reply pkg/kubelet/server/streaming/server_test.go]: I'm fine with it as-is

**Comment by yujuhong** [comment]
> There is no other option.

If tty is set, we get a single stream, but that doesn't mean we have to call it stdout, nor does it mean that we have to assume runtimes can't support multiple streams with tty.

**Comment by ncdc** [comment]
> If tty is set, we get a single stream, but that doesn't mean we have to call it stdout

Fair point, although at least for docker, its api calls them `InputStream`, `OutputStream`, and `ErrorStream`. Do you think it's worth having a different stream in the interface for `ttyOut`?

> , nor does it mean that we have to assume runtimes can't support multiple streams with tty.

Maaaaybe but how would that even work?

**Comment by yujuhong** [comment]
> Fair point, although at least for docker, its api calls them InputStream, OutputStream, and ErrorStream. Do you think it's worth having a different stream in the interface for ttyOut?

That's one option, but I am not sure if it's worth having... @ncdc do you have preferences?

**Comment by ncdc** [comment]
I think keeping it as stdout or renaming to input/output/error makes the
most sense.

**Pull request merged**